man/mono.1
.\"
.\" mono manual page.
.\" Copyright 2003 Ximian, Inc.
.\" Copyright 2004-2011 Novell, Inc.
.\" Copyright 2011-2012 Xamarin Inc
.\" Copyright 2013 7digital Media Ltd.
.\" Author:
.\"   Miguel de Icaza (miguel@gnu.org)
.\"
.TH Mono "mono"
.SH NAME
mono \- Mono's ECMA-CLI native code generator (Just-in-Time and Ahead-of-Time)
.SH SYNOPSIS
.PP
.B mono [options] file [arguments...]
.PP
.B mono-sgen [options] file [arguments...]
.SH DESCRIPTION
\fImono\fP is a runtime implementation of the ECMA Common Language
Infrastructure.  This can be used to run ECMA and .NET applications.
.PP
The runtime loads the specified
.I file
and optionally passes
the
.I arguments
to it.  The
.I file
is an ECMA assembly.  They typically have a .exe or .dll extension.
.PP
These executables can reference additional functionality in the form
of assembly references.  By default those assembly references are
resolved as follows: the \fBmscorlib.dll\fR is resolved from the
system profile that is configured by Mono, and other assemblies are
loaded from the Global Assembly Cache (GAC).
.PP
The runtime contains a native code generator that transforms the
Common Intermediate Language into native code.
.PP
The code generator can operate in two modes: Just-in-time compilation
(JIT) or Ahead-of-time compilation (AOT).  Since code can be
dynamically loaded, the runtime environment and the JIT are always
present, even if code is compiled ahead of time.
.PP
The runtime provides a number of configuration options for running
applications, for developing and debugging, and for testing and
debugging the runtime itself.
.PP
The \fImono\fP command uses the moving and generational SGen garbage collector
while the \fImono-boehm\fP command uses the conservative Boehm
garbage collector.
.SH PORTABILITY
On Unix-based systems, Mono provides a mechanism to emulate the
Windows-style file access, this includes providing a case insensitive
view of the file system, directory separator mapping (from \\ to /) and
stripping the drive letters.
.PP
This functionality is enabled by setting the
.B MONO_IOMAP
environment variable to one of
.B all, drive
and
.B case.
.PP
See the description for
.B MONO_IOMAP
in the environment variables section for more details.
.SH METHOD DESCRIPTIONS
A number of diagnostic command line options take as argument a method
description.   A method description is a textual representation that
can be used to uniquely identify a method.   The syntax is as follows:
.nf
[namespace]classname:methodname[(arguments)]
.fi
.PP
The values in brackets are optional, like the namespace and the
arguments.   The arguments themselves are either empty, or a
comma-separated list of arguments.   Both the
\fBclassname\fR
and
\fBmethodname\fR
can be set to the special value '*' to match any values (Unix shell
users should escape the argument to avoid the shell interpreting
this).
.PP
The arguments, if present should be a comma separated list of types
either a full typename, or for built-in types it should use the
low-level ILAsm type names for the built-in types,
like 'void', 'char', 'bool', 'byte', 'sbyte', 'uint16', 'int16', 'uint',
'int', 'ulong', 'long', 'uintptr', 'intptr', 'single', 'double', 'string', 'object'.
.PP
Pointer types should be the name of the type, followed by a '*',
arrays should be the typename followed by '[' one or more commas (to
indicate the rank of the array), and ']'.
.PP
Generic values should use '<', one or more type names, separated by
both a comma and a space and '>'.
.PP
By-reference arguments should include a "&" after the typename.
.PP
Examples:
.nf
*:ctor(int)		// All constructors that take an int as an argument
*:Main			// Methods named Main in any class
*:Main(string[])	// Methods named Main that take a string array in any class
.fi
.SH RUNTIME OPTIONS
The following options are available:
.TP
\fB--aot\fR, \fB--aot[=options]\fR
This option is used to precompile the CIL code in the specified
assembly to native code.  The generated code is stored in a file with
the extension .so.  This file will be automatically picked up by the
runtime when the assembly is executed.
.Sp
Ahead-of-Time compilation is most useful if you use it in combination
with the -O=all,-shared flag which enables all of the optimizations in
the code generator to be performed.  Some of those optimizations are
not practical for Just-in-Time compilation since they might be very
time consuming.
.Sp
Unlike the .NET Framework, Ahead-of-Time compilation will not generate
domain independent code: it generates the same code that the
Just-in-Time compiler would produce.   Since most applications use a
single domain, this is fine.   If you want to optimize the generated
code for use in multi-domain applications, consider using the
-O=shared flag.
.Sp
This pre-compiles the methods, but the original assembly is still
required to execute as this one contains the metadata and exception
information which is not available on the generated file.  When
precompiling code, you might want to compile with all optimizations
(-O=all).  Pre-compiled code is position independent code.
.Sp
Precompilation is just a mechanism to reduce startup time, increase
code sharing across multiple mono processes and avoid just-in-time
compilation program startup costs.  The original assembly must still
be present, as the metadata is contained there.
.Sp
AOT code typically can not be moved from one computer to another
(CPU-specific optimizations that are detected at runtime) so you
should not try to move the pre-generated assemblies or package the
pre-generated assemblies for deployment.
.Sp
A few options are available as a parameter to the
.B --aot
command line option.   The options are separated by commas, and more
than one can be specified:
.RS
.ne 8
.TP
.I asmonly
Instructs the AOT compiler to output assembly code instead of an
object file.
.TP
.I bind-to-runtime-version
.Sp
If specified, forces the generated AOT files to be bound to the
runtime version of the compiling Mono.   This will prevent the AOT
files from being consumed by a different Mono runtime.
.TP
.I data-outfile=FILE.dll.aotdata
.Sp
This instructs the AOT code generator to output certain data
constructs into a separate file.   This can reduce the executable
images some five to twenty percent.   Developers need to then ship the
resulting aotdata as a resource and register a hook to load the data
on demand by using the
.I mono_install_load_aot_data_hook
method.
.TP
.I direct-icalls
.Sp
When this option is specified, icalls (internal calls made from the
standard library into the mono runtime code) are invoked directly
instead of going through the operating system symbol lookup operation.
This requires use of the
.I static
option.
.TP
.I direct-pinvoke
.Sp
When this option is specified, P/Invoke methods are invoked directly
instead of going through the operating system symbol lookup operation.
This requires use of the
.I static
option.
.TP
.I dwarfdebug
Instructs the AOT compiler to emit DWARF debugging information. When
used together with the nodebug option, only DWARF debugging
information is emitted, but not the information that can be used at
runtime.
.TP
.I full
.Sp
This creates binaries which can be used with the --full-aot option.
.TP
.I hybrid
.Sp
This creates binaries which can be used with the --hybrid-aot option.
.TP
.I llvm
AOT will be performed with the LLVM backend instead of the Mono backend where possible. This will be slower to compile but most likely result in a performance improvement.
.TP
.I llvmonly
AOT will be performed with the LLVM backend exclusively and the Mono backend will not be used. The only output in this mode will be the bitcode file normally specified with the
.I llvm-outfile
option. Use of
.I llvmonly
automatically enables the
.I full
and
.I llvm
options. This feature is experimental.
.TP
.I llvmopts=[options]
Use this option to add more flags to the built-in set of flags passed to the
LLVM optimizer.   When you invoke the
.I mono
command with the
.I --aot=llvm
it displays the current list of flags that are being passed to the
.I opt
command.
.I The list of possible flags that can be passed can be
obtained by calling the bundled
.I opt
program that comes with Mono, and calling it like this:
.nf

	opt --help

.fi

.TP
.I llvmllc=[options]
Use this option to add more flags to the built-in set of flags passed to the
LLVM static compiler (llc).   The list of possible flags that can be passed can be
obtained by calling the bundled
.I llc
program that comes with Mono, and calling it like this:
.nf

	llc --help

.fi
.TP
.I mcpu=[native o generic]
cpu=native allows AOT mode to use all instructions current CPU supports, e.g. AVX2, SSE42, etc.
Default value is 'generic'.
.I mattr=[cpu feature]
Allows AOT code generator to use specified CPU features where possible including `System.Runtime.Intrinsics.*'.
E.g. `mattr=+avx2,mattr=-lzcnt' unlocks sse1-4.2, avx1-2 and disables lzcnt.
It's useful for cross-compilation or when it's not possible to use `-mcpu=native' (which enables
all cpu feature current cpu has).
.I llvm-outfile=[filename]
Gives the path for the temporary LLVM bitcode file created during AOT.
.I dedup
Each AOT module will typically contain the code for inflated methods and wrappers that
are called by code in that module. In dedup mode, we identify and skip compiling all of those
methods. When using this mode with fullaot, dedup-include is required or these methods will
remain missing.
.TP
.I dedup-include=[filename]
In dedup-include mode, we are in the pass of compilation where we compile the methods
that we had previously skipped. All of them are emitted into the assembly that is passed
as this option. We consolidate the many duplicate skipped copies of the same method into one.
.TP
.TP
.I info
Print the architecture the AOT in this copy of Mono targets and quit.
.TP
.I interp
Generates all required wrappers, so that it is possible to run --interpreter without
any code generation at runtime.  This option only makes sense with \fBmscorlib.dll\fR.
Embedders can set
.TP
.I depfile=[filename]
Outputs a gcc -M style dependency file.

.nf
mono_jit_set_aot_mode (MONO_AOT_MODE_INTERP);
.fi
.ne
.TP
.I ld-flags
Additional flags to pass to the C linker (if the current AOT mode calls for invoking it).
.TP
.I llvm-path=<PREFIX>
Same for the llvm tools 'opt' and 'llc'.
.TP
.I msym-dir=<PATH>
Instructs the AOT compiler to generate offline sequence points .msym files.
The generated .msym files will be stored into a subfolder of <PATH> named as the
compilation AOTID.
.TP
.I mtriple=<TRIPLE>
Use the GNU style target triple <TRIPLE> to determine some code generation options, i.e.
--mtriple=armv7-linux-gnueabi will generate code that targets ARMv7. This is currently
only supported by the ARM backend. In LLVM mode, this triple is passed on to the LLVM
llc compiler.
.TP
.I nimt-trampolines=[number]
When compiling in full aot mode, the IMT trampolines must be precreated
in the AOT image.  You can add additional method trampolines with this argument.
Defaults to 512.
.TP
.I ngsharedvt-trampolines=[number]
When compiling in full aot mode, the value type generic sharing trampolines must be precreated
in the AOT image.  You can add additional method trampolines with this argument.
Defaults to 512.
.TP
.I nodebug
Instructs the AOT compiler to not output any debugging information.
.TP
.I no-direct-calls
This prevents the AOT compiler from generating a direct calls to a
method.   The AOT compiler usually generates direct calls for certain
methods that do not require going through the PLT (for example,
methods that are known to not require a hook like a static
constructor) or call into simple internal calls.
.TP
.I nrgctx-trampolines=[number]
When compiling in full aot mode, the generic sharing trampolines must be precreated
in the AOT image.  You can add additional method trampolines with this argument.
Defaults to 4096.
.TP
.I nrgctx-fetch-trampolines=[number]
When compiling in full aot mode, the generic sharing fetch trampolines must be precreated
in the AOT image.  You can add additional method trampolines with this argument.
Defaults to 128.
.TP
.I ntrampolines=[number]
When compiling in full aot mode, the method trampolines must be precreated
in the AOT image.  You can add additional method trampolines with this argument.
Defaults to 4096.
.TP
.I outfile=[filename]
Instructs the AOT compiler to save the output to the specified file.
.TP
.I print-skipped-methods
If the AOT compiler cannot compile a method for any reason, enabling this flag
will output the skipped methods to the console.
.TP
.I profile=[file]
Specify a file to use for profile-guided optimization. See the \fBAOT profiler\fR sub-section. To specify multiple files, include the
.I profile
option multiple times.
.TP
.I profile-only
AOT *only* the methods described in the files specified with the
.I profile
option. See the \fBAOT profiler\fR sub-section.
.TP
.I readonly-value=namespace.typename.fieldname=type/value
Override the value of a static readonly field. Usually, during JIT
compilation, the static constructor is ran eagerly, so the value of
a static readonly field is known at compilation time and the compiler
can do a number of optimizations based on it. During AOT, instead, the static
constructor can't be ran, so this option can be used to set the value of such
a field and enable the same set of optimizations.
Type can be any of i1, i2, i4 for integers of the respective sizes (in bytes).
Note that signed/unsigned numbers do not matter here, just the storage size.
This option can be specified multiple times and it doesn't prevent the static
constructor for the type defining the field to execute with the usual rules
at runtime (hence possibly computing a different value for the field).
.TP
.I save-temps,keep-temps
Instructs the AOT compiler to keep temporary files.
.TP
.I soft-debug
This instructs the compiler to generate sequence point checks that
allow Mono's soft debugger to debug applications even on systems where
it is not possible to set breakpoints or to single step (certain
hardware configurations like the cell phones and video gaming
consoles).
.TP
.I static
Create an ELF object file (.o) or .s file which can be statically linked into an
executable when embedding the mono runtime. When this option is used, the object file
needs to be registered with the embedded runtime using the mono_aot_register_module
function which takes as its argument the mono_aot_module_<ASSEMBLY NAME>_info global
symbol from the object file:

.nf
extern void *mono_aot_module_hello_info;

mono_aot_register_module (mono_aot_module_hello_info);
.fi
.ne
.TP
.I stats
Print various stats collected during AOT compilation.
.TP
.I temp-path=[path]
Explicitly specify path to store temporary files created during AOT compilation.
.TP
.I threads=[number]
This is an experimental option for the AOT compiler to use multiple threads
when compiling the methods.
.TP
.I tool-prefix=<PREFIX>
Prepends <PREFIX> to the name of tools ran by the AOT compiler, i.e. 'as'/'ld'. For
example, --tool=prefix=arm-linux-gnueabi- will make the AOT compiler run
'arm-linux-gnueabi-as' instead of 'as'.
.TP
.I verbose
Prints additional information about type loading failures.
.TP
.I write-symbols,no-write-symbols
Instructs the AOT compiler to emit (or not emit) debug symbol information.
.TP
.I no-opt
Instructs the AOT compiler tot no call opt when compiling with LLVM.
.PP
For more information about AOT, see: http://www.mono-project.com/docs/advanced/aot/
.RE
.TP
\fB--aot-path=PATH\fR
List of additional directories to search for AOT images.
.TP
\fB--apply-bindings=FILE\fR
Apply the assembly bindings from the specified configuration file when running
the AOT compiler.  This is useful when compiling an auxiliary assembly that is
referenced by a main assembly that provides a configuration file.  For example,
if app.exe uses lib.dll then in order to make the assembly bindings from
app.exe.config available when compiling lib.dll ahead of time, use:
.nf
	mono --apply-bindings=app.exe.config --aot lib.dll
.fi
.TP
\fB--assembly-loader=MODE\fR
If mode is \fBstrict\fR, Mono will check that the public key token, culture and version
of a candidate assembly matches the requested strong name.  If mode is \fBlegacy\fR, as
long as the name matches, the candidate will be allowed. \fBstrict\fR is the behavior
consistent with .NET Framework but may break some existing mono-based applications.
The default is \fBlegacy\fR.
.TP
\fB--attach=[options]\fR
Currently the only option supported by this command line argument is
\fBdisable\fR which disables the attach functionality.
.TP
\fB--config filename\fR
Load the specified configuration file instead of the default one(s).
The default files are /etc/mono/config and ~/.mono/config or the file
specified in the MONO_CONFIG environment variable, if set.  See the
mono-config(5) man page for details on the format of this file.
.TP
\fB--debugger-agent=[options]\fR
This instructs the Mono runtime to
start a debugging agent inside the Mono runtime and connect it to a
client user interface will control the Mono process.
This option is typically used by IDEs, like the MonoDevelop or Visual Studio IDEs.
.RS
The configuration is specified using one of more of the following options:
.RS
.ne 8
.TP
.I address=host:port
.Sp
Use this option to specify the IP address where your debugger client is
listening to.
.TP
.I loglevel=LEVEL
.Sp
Specifies the diagnostics log level for
.TP
.I logfile=filename
.Sp
Used to specify the file where the log will be stored, it defaults to
standard output.
.TP
.I server=[y/n]
Defaults to no, with the default option Mono will actively connect to the
host/port configured with the \fBaddress\fR option.  If you set it to 'y', it
instructs the Mono runtime to start debugging in server mode, where Mono
actively waits for the debugger front end to connect to the Mono process.
Mono will print out to stdout the IP address and port where it is listening.
.TP
.I setpgid=[y/n]
If set to yes, Mono will call \fBsetpgid(0, 0)\fR on startup, if that function
is available on the system. This is useful for ensuring that signals delivered
to a process that is executing the debuggee are not propagated to the debuggee,
e.g. when Ctrl-C sends \fBSIGINT\fR to the \fBsdb\fR tool.
.TP
.I suspend=[y/n]
Defaults to yes, with the default option Mono will suspend the vm on startup
until it connects successfully to a debugger front end.  If you set it to 'n', in
conjunction with \fBserver=y\fR, it instructs the Mono runtime to run as normal,
while caching metadata to send to the debugger front end on connection..
.TP
.I transport=transport_name
.Sp
This is used to specify the transport that the debugger will use to
communicate.   It must be specified and currently requires this to
be 'dt_socket'.
.ne
.RE
.TP
\fB--desktop\fR
Configures the virtual machine to be better suited for desktop
applications.  Currently this sets the GC system to avoid expanding
the heap as much as possible at the expense of slowing down garbage
collection a bit.
.TP
\fB--full-aot\fR
This flag instructs the Mono runtime to not
generate any code at runtime and depend exclusively on the code
generated from using mono --aot=full previously.  This is useful for
platforms that do not permit dynamic code generation, or if you need
to run assemblies that have been stripped of IL (for example using
mono-cil-strip).
.Sp
Notice that this feature will abort execution at runtime if a codepath
in your program, or Mono's class libraries attempts to generate code
dynamically.  You should test your software upfront and make sure that
you do not use any dynamic features.
.TP
\fB--full-aot-interp\fR
Same as --full-aot with fallback to the interpreter.
.TP
\fB--gc=boehm\fR, \fB--gc=sgen\fR
Selects the Garbage Collector engine for Mono to use, Boehm or SGen.
Currently this merely ensures that you are running either the
\fImono\fR or \fImono-sgen\fR commands.    This flag can be set in the
\fBMONO_ENV_OPTIONS\fR environment variable to force all of your child
processes to use one particular kind of GC with the Mono runtime.
.TP
\fB--gc-debug=[options]\fR
Command line equivalent of the \fBMONO_GC_DEBUG\fR environment variable.
.TP
\fB--gc-params=[options]\fR
Command line equivalent of the \fBMONO_GC_PARAMS\fR environment variable.
.TP
\fB--arch=32\fR, \fB--arch=64\fR
(Mac OS X only): Selects the bitness of the Mono binary used, if
available. If the binary used is already for the selected bitness, nothing
changes. If not, the execution switches to a binary with the selected
bitness suffix installed side by side (for example, '/bin/mono --arch=64'
will switch to '/bin/mono64' iff '/bin/mono' is a 32-bit build).
.TP
\fB--help\fR, \fB-h\fR
Displays usage instructions.
.TP
\fB--interpreter\fR
The Mono runtime will use its interpreter to execute a given assembly.
The interpreter is usually slower than the JIT, but it can be useful on
platforms where code generation at runtime is not allowed.
.TP
\fB--hybrid-aot\fR
This flag allows the Mono runtime to run assemblies
that have been stripped of IL, for example using mono-cil-strip. For this to
work, the assembly must have been AOT compiled with --aot=hybrid.

This flag is similar to --full-aot, but it does not disable the JIT. This means
you can use dynamic features such as System.Reflection.Emit.
.TP
\fB--llvm\fR
If the Mono runtime has been compiled with LLVM support (not available
in all configurations), Mono will use the LLVM optimization and code
generation engine to JIT or AOT compile.
.Sp
For more information, consult: http://www.mono-project.com/docs/advanced/mono-llvm/
.TP
\fB--ffast-math\fR
This flag allows Mono and LLVM to apply aggressive floating point optimizations.
Can break IEEE754 compliance.
.TP
\fB--nollvm\fR
When using a Mono that has been compiled with LLVM support, it forces
Mono to fallback to its JIT engine and not use the LLVM backend.
.TP
\fB--optimize=MODE\fR, \fB-O=MODE\fR
MODE is a comma separated list of optimizations.  They also allow
optimizations to be turned off by prefixing the optimization name with
a minus sign.
.Sp
In general, Mono has been tuned to use the default set of flags,
before using these flags for a deployment setting, you might want to
actually measure the benefits of using them.
.Sp
The following optimization flags are implemented in the core engine:
.nf
             abcrem     Array bound checks removal
             all        Turn on all optimizations
             aot        Usage of Ahead Of Time compiled code
             branch     Branch optimizations
             cfold      Constant folding
             cmov       Conditional moves [arch-dependency]
             deadce     Dead code elimination
             consprop   Constant propagation
             copyprop   Copy propagation
             fcmov      Fast x86 FP compares [arch-dependency]
             float32	Perform 32-bit float arithmetic using 32-bit operations
             gshared    Enable generic code sharing.
             inline     Inline method calls
             intrins    Intrinsic method implementations
             linears    Linear scan global reg allocation
             leaf       Leaf procedures optimizations
             loop       Loop related optimizations
             peephole   Peephole postpass
             precomp    Precompile all methods before executing Main
             sched      Instruction scheduling
             shared     Emit per-domain code
             sse2       SSE2 instructions on x86 [arch-dependency]
             tailc      Tail recursion and tail calls
.fi
.Sp
For example, to enable all the optimization but dead code
elimination and inlining, you can use:
.nf
	-O=all,-deadce,-inline
.fi
.Sp
The flags that are flagged with [arch-dependency] indicate that the
given option if used in combination with Ahead of Time compilation
(--aot flag) would produce pre-compiled code that will depend on the
current CPU and might not be safely moved to another computer.
.RS
.ne 8
.PP
The following optimizations are supported
.TP
.I float32
Requests that the runtime performn 32-bit floating point operations
using only 32-bits.   By default the Mono runtime tries to use the
highest precision available for floating point operations, but while
this might render better results, the code might run slower.   This
options also affects the code generated by the LLVM backend.
.TP
.I inline
Controls whether the runtime should attempt to inline (the default),
or not inline methods invocations
.ne
.RE
\fB--response=FILE\fR
Provides a response file, this instructs the Mono command to read
other command line options from the specified file, as if the
options had been specified on the command line.   Useful when you have
very long command lines.
.TP
\fB--runtime=VERSION\fR
Mono supports different runtime versions. The version used depends on the program
that is being run or on its configuration file (named program.exe.config). This option
can be used to override such autodetection, by forcing a different runtime version
to be used. Note that this should only be used to select a later compatible runtime
version than the one the program was compiled against. A typical usage is for
running a 1.1 program on a 2.0 version:
.nf
         mono --runtime=v2.0.50727 program.exe
.fi
.TP
\fB--server\fR
Configures the virtual machine to be better suited for server
operations (currently, allows a heavier threadpool initialization).
.TP
\fB--use-map-jit\fR
Instructs Mono to generate code using MAP_JIT on MacOS.  Necessary for
bundled applications.
.TP
\fB--verify-all\fR
Verifies mscorlib and assemblies in the global
assembly cache for valid IL, and all user code for IL
verifiability.

This is different from \fB--security\fR's verifiable
or validil in that these options only check user code and skip
mscorlib and assemblies located on the global assembly cache.
.TP
\fB-V\fR, \fB--version\fR
Prints JIT version information (system configuration, release number
and branch names if available).
.TP
\fB--version=number\fR
Print version number only.


.SH DEVELOPMENT OPTIONS
The following options are used to help when developing a JITed application.
.TP
\fB--debug\fR, \fB--debug=OPTIONS\fR
Turns on the debugging mode in the runtime.  If an assembly was
compiled with debugging information, it will produce line number
information for stack traces.
.RS
.ne 8
.PP
The optional OPTIONS argument is a comma separated list of debugging
options.  These options are turned off by default since they generate
much larger and slower code at runtime.
.TP
The following options are supported:
.TP
.I casts
Produces a detailed error when throwing a InvalidCastException.   This
option needs to be enabled as this generates more verbose code at
execution time.
.TP
.I mdb-optimizations
Disable some JIT optimizations which are usually only disabled when
running inside the debugger.  This can be helpful if you want to attach
to the running process with mdb.
.TP
.I gdb
Generate and register debugging information with gdb. This is only supported on some
platforms, and only when using gdb 7.0 or later.
.ne
.RE
.TP
\fB\-\-profile\fR[=\fIprofiler\fR[:\fIprofiler_args\fR]]\fR
Loads a profiler module with the given arguments. For more information,
see the \fBPROFILING\fR section.
.Sp
This option can be used multiple times; each time will load an additional
profiler module.
.TP
\fB--trace[=expression]\fR
Shows method names as they are invoked.  By default all methods are
traced.
.Sp
The trace can be customized to include or exclude methods, classes or
assemblies.  A trace expression is a comma separated list of targets,
each target can be prefixed with a minus sign to turn off a particular
target.  The words `program', `all' and `disabled' have special
meaning.  `program' refers to the main program being executed, and
`all' means all the method calls.
.Sp
The `disabled' option is used to start up with tracing disabled.  It
can be enabled at a later point in time in the program by sending the
SIGUSR2 signal to the runtime.
.Sp
Assemblies are specified by their name, for example, to trace all
calls in the System assembly, use:
.nf

	mono --trace=System app.exe

.fi
Classes are specified with the T: prefix.  For example, to trace all
calls to the System.String class, use:
.nf

	mono --trace=T:System.String app.exe

.fi
And individual methods are referenced with the M: prefix, and the
standard method notation:
.nf

	mono --trace=M:System.Console:WriteLine app.exe

.fi
Exceptions can also be traced, it will cause a stack trace to be
printed every time an exception of the specified type is thrown.
The exception type can be specified with or without the namespace,
and to trace all exceptions, specify 'all' as the type name.
.nf

	mono --trace=E:System.Exception app.exe

.fi
As previously noted, various rules can be specified at once:
.nf

	mono --trace=T:System.String,T:System.Random app.exe

.fi
You can exclude pieces, the next example traces calls to
System.String except for the System.String:Concat method.
.nf

	mono --trace=T:System.String,-M:System.String:Concat

.fi
You can trace managed to unmanaged transitions using
the wrapper qualifier:
.nf

	mono --trace=wrapper app.exe

.fi
Finally, namespaces can be specified using the N: prefix:
.nf

	mono --trace=N:System.Xml

.fi
.TP
\fB--no-x86-stack-align\fR
Don't align stack frames on the x86 architecture.  By default, Mono
aligns stack frames to 16 bytes on x86, so that local floating point
and SIMD variables can be properly aligned.  This option turns off the
alignment, which usually saves one instruction per call, but might
result in significantly lower floating point and SIMD performance.
.TP
\fB--jitmap\fR
Generate a JIT method map in a /tmp/perf-PID.map file. This file is then
used, for example, by the perf tool included in recent Linux kernels.
Each line in the file has:
.nf

	HEXADDR HEXSIZE methodname

.fi
Currently this option is only supported on Linux.
.SH JIT MAINTAINER OPTIONS
The maintainer options are only used by those developing the runtime
itself, and not typically of interest to runtime users or developers.
.TP
\fB--bisect=optimization:filename\fR
This flag is used by the automatic optimization bug bisector.  It
takes an optimization flag and a filename of a file containing a list
of full method names, one per line.  When it compiles one of the
methods in the file it will use the optimization given, in addition to
the optimizations that are otherwise enabled.  Note that if the
optimization is enabled by default, you should disable it with `-O`,
otherwise it will just apply to every method, whether it's in the file
or not.
.TP
\fB--break method\fR
Inserts a breakpoint before the method whose name is `method'
(namespace.class:methodname).  Use `Main' as method name to insert a
breakpoint on the application's main method.  You can use it also with
generics, for example "System.Collections.Generic.Queue`1:Peek"
.TP
\fB--breakonex\fR
Inserts a breakpoint on exceptions.  This allows you to debug your
application with a native debugger when an exception is thrown.
.TP
\fB--compile name\fR
This compiles a method (namespace.name:methodname), this is used for
testing the compiler performance or to examine the output of the code
generator.
.TP
\fB--compile-all\fR
Compiles all the methods in an assembly.  This is used to test the
compiler performance or to examine the output of the code generator
.TP
\fB--graph=TYPE METHOD\fR
This generates a postscript file with a graph with the details about
the specified method (namespace.name:methodname).  This requires `dot'
and ghostview to be installed (it expects Ghostview to be called
"gv").
.Sp
The following graphs are available:
.nf
          cfg        Control Flow Graph (CFG)
          dtree      Dominator Tree
          code       CFG showing code
          ssa        CFG showing code after SSA translation
          optcode    CFG showing code after IR optimizations
.fi
.Sp
Some graphs will only be available if certain optimizations are turned
on.
.TP
\fB--ncompile\fR
Instruct the runtime on the number of times that the method specified
by --compile (or all the methods if --compile-all is used) to be
compiled.  This is used for testing the code generator performance.
.TP
\fB--stats=[method]\fR
Displays information about the work done by the runtime during the

packaging/MacSDK/mono.py
import os
import re

from bockbuild.package import Package
from bockbuild.util.util import *


class MonoMasterPackage(Package):
    def __init__(self):
        Package.__init__(
            self,
            "mono",
            None,
            sources=[Package.profile.git_root],
            git_branch=os.getenv("MONO_BRANCH") or None,
            revision=os.getenv("MONO_BUILD_REVISION"),
            configure_flags=["--enable-nls=no", "--with-ikvm=yes"],
        )
        self.source_dir_name = "mono"
        # This package would like to be lipoed.
        self.needs_lipo = True

        # Don't clean the workspace, so we can run 'make check' afterwards
        self.dont_clean = True

        if Package.profile.name == "darwin":
            self.configure_flags.extend([
                "--with-libgdiplus=%s/lib/libgdiplus.dylib" %
                Package.profile.staged_prefix,
                "--enable-llvm",
                "CXXFLAGS=-stdlib=libc++",
            ])

            self.sources.extend([
                # Fixes up pkg-config usage on the Mac
                "patches/mcs-pkgconfig.patch"
            ])
        else:
            self.configure_flags.extend([
                "--with-libgdiplus=%s/lib/libgdiplus.so" %
                Package.profile.staged_prefix
            ])

        self.gcc_flags.extend(["-O2"])

        self.configure = './autogen.sh --prefix="%{package_prefix}"'

        self.extra_stage_files = ["etc/mono/config"]
        self.custom_version_str = None

    def build(self):
        self.make = "%s EXTERNAL_RUNTIME=%s" % (self.make,
                                                self.profile.env.system_mono)
        Package.configure(self)

        if self.custom_version_str is not None:
            replace_in_file(
                os.path.join(self.workspace, "config.h"),
                {self.version: self.custom_version_str},
            )
        Package.make(self)

    def prep(self):
        Package.prep(self)
        for p in range(1, len(self.local_sources)):
            self.sh('patch -p1 < "%{local_sources[' + str(p) + ']}"')

    def arch_build(self, arch):
        Package.profile.arch_build(arch, self)
        if arch == "darwin-64":  # 64-bit build pass
            self.local_configure_flags.extend(
                ["--build=x86_64-apple-darwin13.0.0", "--disable-boehm"])

        if arch == "darwin-32":  # 32-bit build pass
            self.local_configure_flags.extend(
                ["--build=i386-apple-darwin13.0.0"])

        self.local_configure_flags.extend([
            "--cache-file=%s/%s-%s.cache" %
            (self.profile.bockbuild.build_root, self.name, arch)
        ])

    def install(self):
        Package.install(self)

        registry_dir = os.path.join(self.staged_prefix, "etc", "mono",
                                    "registry", "LocalMachine")
        ensure_dir(registry_dir)

        # LLVM build installs itself under the source tree; move tools to mono's install path
        llvm_tools_path = os.path.join(self.workspace, "llvm/usr/bin")
        target = os.path.join(self.staged_prefix, "bin")
        ensure_dir(target)
        for tool in ["opt", "llc"]:
            shutil.move(os.path.join(llvm_tools_path, tool), target)

    def deploy(self):
        if bockbuild.cmd_options.arch == "darwin-universal":
            os.symlink("mono-sgen64", "%s/bin/mono64" % self.staged_profile)
            os.symlink("mono-sgen32", "%s/bin/mono32" % self.staged_profile)


MonoMasterPackage()
